python-checks/src/main/java/org/sonar/python/checks/BuiltinShadowingAssignmentCheck.java

```diff
@@ -0,0 +1,156 @@
+  }
+
+  private void checkAssignmentExpression(SubscriptionContext ctx) {
+    AssignmentExpression assignmentExpression = (AssignmentExpression) ctx.syntaxNode();
+    if (shouldReportIssue(assignmentExpression.lhsName(), assignmentExpression.lhsName())) {
+      ctx.addIssue(assignmentExpression.lhsName(), String.format(MESSAGE, VRBL_ISSUE_TYPE));
+    }
+  }
+
+  private void checkAssignment(SubscriptionContext ctx) {
+    AssignmentStatement assignment = (AssignmentStatement) ctx.syntaxNode();
+    for (int i = 0; i < assignment.lhsExpressions().size(); i++) {
+      for (Expression expression : assignment.lhsExpressions().get(i).expressions()) {
+        if (shouldReportIssue(expression, expression)) {
+          ctx.addIssue(expression, String.format(MESSAGE, VRBL_ISSUE_TYPE));
+        }
+      }
+    }
+  }
+
+  private void checkAnnotatedAssignment(SubscriptionContext ctx) {
+    AnnotatedAssignment assignment = (AnnotatedAssignment) ctx.syntaxNode();
+    Expression variable = assignment.variable();
+    Token equalToken = assignment.equalToken();
+    if (equalToken != null && shouldReportIssue(variable, variable)) {
+      ctx.addIssue(variable, String.format(MESSAGE, VRBL_ISSUE_TYPE));
+    }
+  }
+
+  private void checkFunctionDeclaration(SubscriptionContext ctx) {
+    FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+    Tree ancestor = TreeUtils.firstAncestorOfKind(functionDef, Tree.Kind.FUNCDEF, Tree.Kind.CLASSDEF);
+    // We should not raise the issue in case of a method
+    if ((ancestor == null || ancestor.getKind().equals(Tree.Kind.FUNCDEF)) && shouldReportIssue(functionDef.name(), functionDef)) {
+      ctx.addIssue(functionDef.name(), String.format(MESSAGE, "function"));
+    }
+    ParameterList parameterList = functionDef.parameters();
+    if (parameterList != null) {
+      for (Parameter parameter : parameterList.nonTuple()) {
+        Name parameterName = parameter.name();
+        if (parameterName != null && isBuiltInName(parameterName)) {
+          ctx.addIssue(parameterName, String.format(MESSAGE, "parameter"));
+        }
+      }
+    }
+  }
+
+  private void checkClassDefinition(SubscriptionContext ctx) {
+    ClassDef classDef = (ClassDef) ctx.syntaxNode();
+    if (shouldReportIssue(classDef.name(), classDef)) {
+      ctx.addIssue(classDef.name(), String.format(MESSAGE, "class"));
+    }
+  }
+
+  private void checkImportedNameAlias(SubscriptionContext ctx, AliasedName aliasedName) {
+    Name alias = aliasedName.alias();
+    if (alias != null && shouldReportIssue(alias, aliasedName)) {
+      ctx.addIssue(alias, String.format(MESSAGE, "alias"));
+    }
+  }
+
+  private boolean shouldReportIssue(Tree tree, Tree parent) {
+    return tree.is(Tree.Kind.NAME) && isBuiltInName((Name) tree) && !skipIssueCondition(parent);
+  }
+
+  private boolean isBuiltInName(Name name) {
+    return reservedNames.contains(name.name());
+  }
+
+  private static boolean skipIssueCondition(Tree name) {
```

Comment:
_minor_ I'd rename this method into something that better reflects the fact that it returns a boolean (like `isException`, `shouldSkipIssueCondition`, or something like that)